llumnix/backends/vllm/migrate_backend.py

```diff
@@ -0,0 +1,257 @@
+        dummy_value_cpu = data[num_blocks:2*num_blocks]
+        with torch.cuda.stream(self.migration_stream):
+            for layer_idx in range(self.cache_engine.num_layers):
+                for idx, block_num in enumerate(blocks):
+                    dummy_key_cpu[idx][layer_idx].copy_(self.gpu_cache[layer_idx][0][block_num], non_blocking=True)
+                    dummy_value_cpu[idx][layer_idx].copy_(self.gpu_cache[layer_idx][1][block_num], non_blocking=True)
+        torch.cuda.Stream.synchronize(self.migration_stream)
+        col.send(data, dst_handle, self.group_name)
+
+    def do_recv(self, src_handle, blocks: List[int]):
+        num_blocks = len(blocks)
+        data = self.dummy_cache[:2*num_blocks]
+        # note that col.recv use ray.collective inner stream, not migration_stream
+        col.recv(data, src_handle, self.group_name)
+
+        dummy_key_cpu = data[:num_blocks]
+        dummy_value_cpu = data[num_blocks:2*num_blocks]
+        with torch.cuda.stream(self.migration_stream):
+            for layer_idx in range(self.cache_engine.num_layers):
+                for idx, block_num in enumerate(blocks):
+                    self.gpu_cache[layer_idx][0][block_num].copy_(dummy_key_cpu[idx][layer_idx], non_blocking=True)
+                    self.gpu_cache[layer_idx][1][block_num].copy_(dummy_value_cpu[idx][layer_idx], non_blocking=True)
+        torch.cuda.Stream.synchronize(self.migration_stream)
+
+def get_migrate_backend(migrate_config: MigrationConfig, cache_engine: CacheEngine, worker_handle_list, scheduling_strategy, \
+                           dtype, is_driver_worker, gpu_cache, ray_world_size, ray_rank, worker_rank, local_rank) -> MigrateBackendBase:
+    if migrate_config.pp_or_tp_enabled and migrate_config.migration_backend == 'nccl':
+        logger.warning("NCCL backend is not supported for PP or TP enabled model, using gloo instead.")
+        migrate_config.migration_backend = 'gloo'
+
+    if cache_engine.num_gpu_blocks < migrate_config.migration_cache_blocks:
+        logger.warning("migration_cache_blocks({}) is larger than num_gpu_blocks({}), reducing it to num_gpu_blocks."
+                       .format(migrate_config.migration_cache_blocks, cache_engine.num_gpu_blocks))
+        migrate_config.migration_cache_blocks = cache_engine.num_gpu_blocks
+
+    target_col = None
+    backend = migrate_config.migration_backend
+    if backend in ['nccl', 'gloo']:
+        target_col = RayMigrateBackend(migrate_config, cache_engine, ray_world_size, ray_rank, \
+                                          local_rank, scheduling_strategy, dtype, is_driver_worker, gpu_cache)
+    elif backend == 'rpc':
+        target_col = RPCMigrateBackend(migrate_config, cache_engine, worker_rank, worker_handle_list, scheduling_strategy, \
+                                           dtype, is_driver_worker, gpu_cache)
+    else:
+        raise ValueError(f"Unsupported backend {backend}")
```

Comment:
There is choices option in migration backend argument, so it is unnecessary to detect here.